File: LeetCode/200-number-of-islands.py
###############################################################################################
# disjoint set union
###########
# Time Complexity: O(n^2)
# Space Complexity: O(n^2)
###############################################################################################
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        f = [[(i,j) for j in range(len(grid[0]))] for i in range(len(grid))]
        def find(lo):
            i, j = lo[0], lo[1]
            if f[i][j] != (i,j):
                f[i][j] = find(f[i][j])
            return f[i][j]
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    if i-1 >= 0 and grid[i-1][j] == '1':
                        x,y = find((i,j))
                        f[x][y] = find((i-1,j))
                    if j-1 >= 0 and grid[i][j-1] == '1':
                        x,y = find((i,j))
                        f[x][y] = find((i,j-1))
                    if i+1 < len(grid) and grid[i+1][j] == '1':
                        x,y = find((i,j))
                        f[x][y] = find((i+1,j))
                    if j+1 < len(grid[0]) and grid[i][j+1] == '1':
                        x,y = find((i,j))
                        f[x][y] = find((i,j+1))
        res = set()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    res.add(find((i, j)))
        return len(res)


###############################################################################################
# dfs, this is 1x faster than above
###########
# Time Complexity: O(n^2)
# Space Complexity: O(n^2), the worst is n^2, but usually we won't achieve that
###############################################################################################
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(i, j):
            grid[i][j] = '0'
            if i+1 < len(grid) and grid[i+1][j] == '1':
                dfs(i+1, j)
            if j+1 < len(grid[0]) and grid[i][j+1] == '1':
                dfs(i, j+1)
            if i-1 >= 0 and grid[i-1][j] == '1':
                dfs(i-1, j)
            if j-1 >= 0 and grid[i][j-1] == '1':
                dfs(i, j-1)
        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    res += 1
                    dfs(i,j)
        return res

File: LeetCode/200-number-of-islands.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// easy
////////////////
// Time Complexity: O(m*n)
// Space Complexity: O(m*n)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    int dirs[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    int numIslands(vector<vector<char>>& grid) {
        int res = 0;
        for (int i=0 ; i<grid.size() ; i++){
            for (int j=0 ; j<grid[0].size() ; j++){
                if (grid[i][j] == '1') {
                    res += 1;
                    dfs(grid, i, j);
                }
            }
        }
        return res;
    }
    void dfs(vector<vector<char>>& grid, int i, int j){
        grid[i][j] = '0';
        for (auto dir:dirs){
            int d0=dir[0], d1=dir[1];
            if (i+d0 < grid.size() && j+d1 < grid[0].size() && grid[i+d0][j+d1] == '1')
                dfs(grid, i+d0, j+d1);
        }
    }
};